quint/src/flattening.ts

```diff
@@ -72,349 +64,412 @@ export function flattenModules(
 
-  const context = { idGenerator, table, currentModuleNames, sourceMap, analysisOutput, importedModules }
+  const flattenedDefs = flattener.flattenDef(def)
+  const flattenedModule: FlatModule = { ...module, defs: [...module.defs, ...flattenedDefs] }
 
-  const newDefs = module.defs.flatMap(def => {
+  return {
+    flattenedModule,
+    flattenedTable: resolveNamesOrThrow(table, sourceMap, flattenedModule),
+    flattenedAnalysis: analysisOutput,
+  }
+}
+
+class Flatenner {
+  private idGenerator: IdGenerator
+  private table: LookupTable
+  private currentModuleNames: Set<string>
+  private sourceMap: Map<bigint, Loc>
+  private analysisOutput: AnalysisOutput
+  private importedModules: Map<string, QuintModule>
+  private module: QuintModule
+
+  constructor(
+    idGenerator: IdGenerator,
+    table: LookupTable,
+    sourceMap: Map<bigint, Loc>,
+    analysisOutput: AnalysisOutput,
+    importedModules: Map<string, QuintModule>,
+    module: QuintModule
+  ) {
+    this.idGenerator = idGenerator
+    this.table = table
+    this.currentModuleNames = new Set([
+      // builtin names
+      ...defaultValueDefinitions().map(d => d.identifier),
+      // names from the current module
+      ...compact(module.defs.map(d => (isFlat(d) ? d.name : undefined))),
+    ])
+
+    this.sourceMap = sourceMap
+    this.analysisOutput = analysisOutput
+    this.importedModules = importedModules
+    this.module = module
+  }
+
+  flattenDef(def: QuintDef): FlatDef[] {
     if (isFlat(def)) {
       // Not an instance, import or export, keep the same def
       return [def]
     }
 
     if (def.kind === 'instance') {
-      return flattenInstance(context, def)
+      return this.flattenInstance(def)
     }
 
-    return flattenImportOrExport(context, def)
-  })
+    return this.flattenImportOrExport(def)
+  }
 
-  return { ...module, defs: uniqBy(newDefs, 'name') }
-}
+  flattenModule(): FlatModule {
+    const newDefs = this.module.defs.flatMap(def => this.flattenDef(def))
 
-function isFlat(def: QuintDef): def is FlatDef {
-  return def.kind !== 'instance' && def.kind !== 'import' && def.kind !== 'export'
-}
+    return { ...this.module, defs: uniqBy(newDefs, 'name') }
+  }
 
-function flattenInstance(context: FlatteningContext, def: QuintInstance): FlatDef[] {
-  // Build pure val definitions from overrides to replace the constants in the
-  // instance. Index them by name to make it easier to replace the corresponding constants.
-  const overrides: Map<string, FlatDef> = new Map(
-    def.overrides.map(([param, expr]) => {
-      const constDef = context.table.get(param.id)!
-
-      return [
-        param.name,
-        {
-          kind: 'def',
-          qualifier: 'pureval',
-          name: param.name,
-          expr,
-          typeAnnotation: constDef.typeAnnotation,
-          id: param.id,
-        },
-      ]
-    })
-  )
+  private flattenInstance(def: QuintInstance): FlatDef[] {
```

Comment:
This file from here on is pretty much unchanged. The only change is that we now use `this` instead of a `ctx` parameter`.